ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB.hs

```diff
@@ -1,82 +1,1987 @@
+--
+-- - __Maintaining the in-memory ledger state at the tip__: Maintaining the
+--     ledger state corresponding to the current tip in memory. When we try to
+--     extend our chain with a new block fitting onto our tip, the block must
+--     first be validated using the right ledger state, i.e., the ledger state
+--     corresponding to the tip. The current ledger state is needed for various
+--     other purposes.
+--
+-- - __Maintaining the past \(k\) in-memory ledger states__: we might roll back
+--     up to \(k\) blocks when switching to a more preferable fork. Consider the
+--     example below:
+--
+--     <<docs/haddocks/ledgerdb-switch.svg>>
+--
+--     Our current chain's tip is \(C_2\), but the fork containing blocks
+--     \(F_1\), \(F_2\), and \(F_3\) is more preferable. We roll back our chain
+--     to the intersection point of the two chains, \(I\), which must be not
+--     more than \(k\) blocks back from our current tip. Next, we must validate
+--     block \(F_1\) using the ledger state at block \(I\), after which we can
+--     validate \(F_2\) using the resulting ledger state, and so on.
+--
+--     This means that we need access to all ledger states of the past \(k\)
+--     blocks, i.e., the ledger states corresponding to the volatile part of the
+--     current chain. Note that applying a block to a ledger state is not an
+--     invertible operation, so it is not possible to simply /unapply/ \(C_1\)
+--     and \(C_2\) to obtain \(I\).
+--
+--     Access to the last \(k\) ledger states is not only needed for validating
+--     candidate chains, but also by the:
+--
+--     - __Local state query server__: To query any of the past \(k\) ledger
+--       states.
+--
+--     - __Chain sync client__: To validate headers of a chain that intersects
+--        with any of the past \(k\) blocks.
+--
+-- - __Provide 'Ouroboros.Consensus.Ledger.Tables.Basics.LedgerTable's at any of the last \(k\) ledger states__: In order to apply blocks or transactions
+--     on top of ledger states, the LedgerDB must be able to provide the
+--     appropriate ledger tables at any of those ledger states.
+--
+-- - __Storing on disk__: To obtain a ledger state for the current tip of the
+--     chain, one has to apply /all blocks in the chain/ one-by-one to the
+--     initial ledger state. When starting up the system with an on-disk chain
+--     containing millions of blocks, all of them would have to be read from
+--     disk and applied. This process can take hours, depending on the storage
+--     and CPU speed, and is thus too costly to perform on each startup.
+--
+--     For this reason, a recent snapshot of the ledger state should be
+--     periodically written to disk. Upon the next startup, that snapshot can be
+--     read and used to restore the current ledger state, as well as the past
+--     \(k\) ledger states.
+--
+-- - __Flushing 'LedgerTable' differences__: The running Consensus has to
```

Comment:
Would it be possible to link to a module where differences are explained (and possibly defined)?